cocotb/share/lib/embed/embed.cpp

```diff
@@ -0,0 +1,139 @@
+// Copyright cocotb contributors
+// Licensed under the Revised BSD License, see LICENSE for details.
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <embed.h>
+#include <cocotb_utils.h>   // xstr, utils_dyn_open, utils_dyn_sym
+#include <gpi.h>            // gpi_event_t
+#include <cstdlib>          // getenv
+#if ! defined(__linux__) && ! defined(__APPLE__)
+#include <windows.h>        // Win32 API for loading the embed impl library
+#include <string>           // string
+#endif
+
+#ifndef EMBED_IMPL_LIB
+#error "Name of embed implementation library required"
+#else
+#define EMBED_IMPL_LIB_STR xstr(EMBED_IMPL_LIB)
+#endif
+
+
+static void (*_embed_init_python)();
+static void (*_embed_sim_cleanup)();
+static int (*_embed_sim_init)(int argc, char const * const * argv);
+static void (*_embed_sim_event)(gpi_event_t level, const char *msg);
+
+static bool init_failed = false;
+
+#if ! defined(__linux__) && ! defined(__APPLE__)
+HANDLE act_ctx = NULL;
+
+BOOL WINAPI DllMain(HINSTANCE, DWORD fdwReason, LPVOID)
+{
+
+    if (fdwReason == DLL_PROCESS_ATTACH) {
+        // Save current activation context.
+        GetCurrentActCtx(&act_ctx);
+    } else if (fdwReason == DLL_PROCESS_DETACH) {
+        if (act_ctx) {
+            ReleaseActCtx(act_ctx);
+        }
+    }
+
+    return TRUE;
+}
+#endif
+
+static void * load_embed_impl_lib()
+{
+    return utils_dyn_open(EMBED_IMPL_LIB_STR);
+}
+
+
+extern "C" void embed_init_python(void)
+{
+    // preload python library
+    auto libpython_path = getenv("LIBPYTHON_LOC");
+    if (!libpython_path) {
+        LOG_ERROR("Environment variable LIBPYTHON_LOC not set to the path of your environment's libpython");
```

Comment:
Sorry for going this one more time. I did some basic testing and it looks `cocotb-config` can be called even on currently broken xcelium. Do not have a strong opinion but what you think to try to call it in case  `LIBPYTHON_LOC` not set with some worning?

Reply:
Are you suggesting calling `cocotb-config --libpython` from embed.cpp? If this is desired then we should move the call unconditionally to embed.cpp. I'm just afraid that won't work in certain cases if `cocotb-config` is not on the PATH; for instance on Windows when you are not using conda, or maybe a simulator modifies PATH.

I'd rather it always work without fail and have 1 additional requirement on advanced users, than ___usually___ work and save little work. I had thought it would be a good idea to try defaulting to "libpythonX.X.so" if `LIBPYTHON_LOC` is not set and that will work in the common case without setting `LIBPYTHON_LOC`.

Reply:
I added the `libpythonX.X.so` default back in.

Reply:
In principle, we require `cocotb-config` to be on the `PATH` always: https://docs.cocotb.org/en/stable/install.html#installation-of-cocotb
Like now is also ok. As I said no strong opinion. I Will test this a bit in next days.

Reply:
Thank you for testing it.

Reply:
All platform and simulators I have access seem to be happy.